File: packages/i18nify-js/src/modules/banking/normalizeIFSC.ts
import { withErrorBoundary } from '../../common/errorBoundary';

const IFSC_REGEX = /^[A-Z]{4}0[A-Z0-9]{6}$/;

const normalizeIFSC = (ifsc: string): string => {
  if (!ifsc || typeof ifsc !== 'string')
    throw new Error('IFSC must be a non-empty string.');

  const normalised = ifsc.trim().toUpperCase();

  if (!IFSC_REGEX.test(normalised))
    throw new Error(
      `Invalid IFSC format: "${ifsc}". Expected 4 letters + '0' + 6 alphanumeric characters (e.g. SBIN0001234).`,
    );

  return normalised;
};

export default withErrorBoundary<typeof normalizeIFSC>(normalizeIFSC);

File: packages/i18nify-go/modules/banking/normalize_ifsc.go
package banking

import (
	"fmt"
	"strings"
)

// NormalizeIFSC upper-cases and trims ifsc, returning an error if the result
// does not match the standard IFSC format (4 letters + '0' + 6 alphanumeric).
func NormalizeIFSC(ifsc string) (string, error) {
	if strings.TrimSpace(ifsc) == "" {
		return "", fmt.Errorf("banking: NormalizeIFSC: ifsc must be a non-empty string")
	}
	n := strings.ToUpper(strings.TrimSpace(ifsc))
	if !ifscRegex.MatchString(n) {
		return "", fmt.Errorf(
			"banking: NormalizeIFSC: invalid IFSC format %q — expected 4 letters + '0' + 6 alphanumeric (e.g. SBIN0001234)",
			ifsc,
		)
	}
	return n, nil
}
